### ISSUE TITLE
Using multiple optimizers for the same architecture

### ISSUE THREAD

**Issue by CuauLuzbel** [PyTorchLightning/pytorch-lightning#3346]
Is there a way to use two different optimizers for different parts of the architecture without repeating the training step twice?

**Comment by awaelchli** [comment]
https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html#configure-optimizers

no, but we typically need to do different things for each optimizer, therefore we need to call training_step multiple times with a different optimizer_idx as argument. 

If you can provide some pseudo code, I may better understand what you want to do and I can try help convert it to Lightning.

**Comment by awaelchli** [comment]
@CuauLuzbel did I answer your question? Let me know if anything is unclear

**Comment by soumyasanyal** [comment]
Hi @awaelchli,

My requirements are similar to this issue and also [this one](https://github.com/PyTorchLightning/pytorch-lightning/issues/3742). The linked issue already has the pseudo-code. It would help a lot of us if you could share how to solve this in the lightning framework.

**Comment by awaelchli** [comment]
First, check out the [manual optimization docs](https://pytorch-lightning.readthedocs.io/en/stable/optimizers.html#manual-optimization), there you will already find a GAN example.
There you have two parts, a discriminator and a generator, just as in the other post you mention. 
You can call step and manual backward as you wish. It's almost identical to what you would do in pure pytorch.

**Comment by awaelchli** [comment]
My interpretation of what the other poster wanted to do:

```python
@property
def automatic_optimization(self):
    return False

def training_step(self, batch, batch_idx):
    out = self.first_module(batch)
    out = self.second_module(out)

    opt1, opt2 = self.optimizes()

    loss = criterion(out, batch['target'])

    self.manual_backward(loss)
    opt1.step()
    opt2.step()


def configure_optimizers(self):
    optimizer_first_module = torch.optim.Adam(self.first_module.parameters())
    optimizer_second_module = torch.optim.Adam(self.second_module.parameters())
    return [optimizer_first_module, optimizer_second_module]
```

I didn't run this code, there may be a typo or two. I hope it makes a bit more sense?

**Comment by soumyasanyal** [comment]
Thanks @awaelchli, this looks very promising. There is one small issue though. Currently, calling just `self.manual_backward(loss)` is not allowed within lightning as we need to pass an optimizer as well (required arg), i.e., call `self.manual_backward(loss, opt1)` instead. I checked the code for manual_backward - it uses the optimizer param to clip the gradients and synchronize the optimizer (during DDP I think).

From the docs, one way to override the backward call is to redefine the `LightningModule.backward()` function to just perform `loss.backward()` and ignore any optimizers passed. That'll override the optimizer specific code being run within `manual_backward`. But I'd love to hear your thoughts as I'm quite new to the framework. Is this the cleanest solution to the problem?

**Comment by awaelchli** [comment]
same question asked on slack: https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1610233885334100
Colab demo with a possible solution:
https://colab.research.google.com/drive/1jVPI6as9gBCRxdu7r1Q6RvYu2Jh08OKJ?usp=sharing

**Comment by pasinit** [comment]
my issue is slightly different, I simply have two parts of the same network trained with two different learning rates, e.g., first K layers with one LR and last N layers with another one. However the input and the loss I would use for the backward pass would be the exact same, so it is a waste of computation to pass the same input through the network twice. Can this scenario be implemented in lightning or I need to resort to the manual training method? Thanks :)

**Comment by awaelchli** [comment]
@pasinit In PyTorch, you can do one optimizer with learning rates per parameter groups: https://pytorch.org/docs/stable/optim.html#per-parameter-options

I would go this way unless there are more specific reasons for needing multiple optimizers in this use case.

**Comment by vanderschuea** [comment]
> same question asked on slack: https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1610233885334100 Colab demo with a possible solution: https://colab.research.google.com/drive/1jVPI6as9gBCRxdu7r1Q6RvYu2Jh08OKJ?usp=sharing

While this solution is awesome, and clearly worked at the time, this isn't the case anymore. Pytorch-lightning has changed the optimizer-step to now work with closures and putting the call to training_step inside the closure. Failing to call said closure incurs an error: `MisconfigurationException: The closure hasn't been executed`.

There are now still 2 workarounds :

1. Write your own custom training loop inside `training_step` by setting `self.automatic_optimization` to `False`.
2. Wrap the 2 optimizers inside one new 'hybrid' optimizer (this also necessitates a wrapper around lr_scheduler if you want to apply a different lr_scheduler to each optimizer)

I prefer solution 2 since it avoids supplementary coding to support all the pytorch-lightning parameters like gradient accumulation. It however asks to add 2 classes to our code like so (this code was tested and works even when resuming training from a checkpoint):

```
import torch
from collections import defaultdict
class HybridOptim(torch.optim.Optimizer):
    # Wrapper around multiple optimizers that should be executed at the same time
    def __init__(self, optimizers):
        self.optimizers = optimizers

    @property
    def state(self):
        state = defaultdict(dict)
        for optimizer in self.optimizers:
            state = {**state, **optimizer.state}
        return state

    @property
    def param_groups(self):
        param_groups = []
        for optimizer in self.optimizers:
            param_groups = param_groups + optimizer.param_groups
        return param_groups

    def __getstate__(self):
        return [optimizer.__getstate__() for optimizer in self.optimizers]

    def __setstate__(self, state):
        for opt_state, optimizer in zip(self.optimizers, state):
            optimizer.__setstate__(opt_state)

    def __repr__(self):
        format_string = self.__class__.__name__ + ' ('
        for optimizer in self.optimizers:
            format_string += '\n'
            format_string += optimizer.__repr__()
        format_string += ')'
        return format_string

    def _hook_for_profile(self):
        for optimizer in self.optimizers:
            optimizer._hook_for_profile()

    def state_dict(self):
        return [optimizer.state_dict() for optimizer in self.optimizers]

    def load_state_dict(self, state_dict):
        for state, optimizer in zip(state_dict, self.optimizers):
            optimizer.load_state_dict(state)

    def zero_grad(self, set_to_none: bool = False):
        for optimizer in self.optimizers:
            optimizer.zero_grad(set_to_none=set_to_none)

    def add_param_group(self, param_group):
        raise NotImplementedError()

    def step(self, closure=None):
        loss = None
        if closure is not None:
            with torch.enable_grad():
                loss = closure()

        for optimizer in self.optimizers:
            optimizer.step()

        return loss

class HybridLRS():
    """ Wrapper Class around lr_scheduler to return a 'dummy' optimizer to pass
        pytorch lightning checks
    """
    def __init__(self, hybrid_optimizer, idx, lr_scheduler) -> None:
        self.optimizer = hybrid_optimizer
        self.idx = idx
        self.lr_scheduler = lr_scheduler

    def __getattribute__(self, __name: str):
        if __name in {"optimizer", "idx", "lr_scheduler"}:
            return super().__getattribute__(__name)
        else:
            return self.lr_scheduler.__getattribute__(__name)

class PLModule():
    def configure_optimizers(self):
        # Classic initialization
        optimizer1 = ...
        optimizer2 = ...
        lrs1 = LRS(optimizer1, ...)
        lrs2 = LRS(optimizer2, ...)
        hoptim = HybridOptim([opt1, opt2])
        return [hoptim], [HybridLRS(hoptim, 0, lrs1), HybridLRS(hoptim, 1, lrs2)]
```
Hope this can help people who also struggled with this issue :) (this code should not break even during future pytorch-lightning updates since it works independently of any specifics of the framework)

**Comment by jamesnordlund** [comment]
@vanderschuea very nice solution, I appreciate you sharing it here.  Sorry to bump an old thread like this, but it seems version 1.9.0 kills your approach.  I get the error:
> lightning_fabric.utilities.exceptions.MisconfigurationException: Unknown configuration for model optimizers.

Rolling back to 1.8.6 gets my code to run as expected.  

Wanted to comment an update here because AFAIK this is still the most useful thread online about how to use different optimizers in one architecture with Lightning.

**Comment by nathancooperjones** [comment]
Updating the code to include a `defaults` property seems to fix the issue in PyTorch Lightning 2.0. I'll post my working code below: 

```python
class HybridOptim(torch.optim.Optimizer):
    """
    Wrapper around multiple optimizers that should be stepped together at a single time. This is
    a hack to avoid PyTorch Lightning calling ``training_step`` once for each optimizer, which
    increases training time and is not always necessary.

    Modified from the reply in a GitHub Issue thread here:
    https://github.com/Lightning-AI/lightning/issues/3346#issuecomment-1036063687

    Parameters
    ----------
    optimizers: list of optimizers

    """
    def __init__(self, optimizers: Iterable[torch.optim.Optimizer]) -> None:
        self.optimizers = optimizers

    @property
    def state(self) -> Dict[str, torch.Tensor]:
        """Return the combined state for each optimizer in ``self.optimizers``."""
        return {
            key: value
            for optimizer in self.optimizers
            for key, value in optimizer.state.items()
        }

    @property
    def param_groups(self) -> List[Dict[str, Union[torch.Tensor, float, bool, Any]]]:
        """Return the combined parameter groups for each optimizer in ``self.optimizers``."""
        return [
            element
            for optimizer in self.optimizers
            for element in optimizer.param_groups
        ]

    # !!!!!! HERE IS THE NEW CODE !!!!!!!!!
    @property
    def defaults(self) -> Dict[str, torch.Tensor]:
        """Return the combined defaults for each optimizer in ``self.optimizers``."""
        return {
            key: value
            for optimizer in self.optimizers
            for key, value in optimizer.defaults.items()
        }
    # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

    def __getstate__(self) -> List[torch.optim.Optimizer]:
        """Return ``self.optimizers`` for pickling purposes."""
        return self.optimizers

    def __setstate__(self, optimizers: List[torch.optim.Optimizer]) -> None:
        """
        Load ``optimizers`` into ``self.optimizers`` for pickling purposes and call
        ``__setstate__``.

        """
        self.optimizers = optimizers

        # call remaining lines of the ``torch.optim.Optimizer.__setstate__`` method just to be safe.
        # copied from: https://pytorch.org/docs/stable/_modules/torch/optim/optimizer.html#Optimizer
        for optimizer in self.optimizers:
            optimizer._hook_for_profile()  # To support multiprocessing pickle/unpickle.
            optimizer.defaults.setdefault('differentiable', False)

    def __repr__(self) -> str:
        """Call and concatenate ``__repr__`` for each optimizer in ``self.optimizers``."""
        repr_str = f'``{self.__class__.__name__}`` containing {len(self.optimizers)} optimizers:\n'

        for optimizer in self.optimizers:
            repr_str += '\n' + optimizer.__repr__()

        return repr_str

    def _hook_for_profile(self) -> None:
        """Call ``_hook_for_profile`` for each optimizer in ``self.optimizers``."""
        for optimizer in self.optimizers:
            optimizer._hook_for_profile()

    def state_dict(self) -> List[
        Dict[str, Union[torch.Tensor, List[Dict[str, Union[torch.Tensor, float, bool, Any]]]]]
    ]:
        """
        Returns the state of the optimizer as a dictionary.

        It contains two entries:

            * ``state`` - a dict holding current optimization state. Its content differs between
              optimizer classes.
            * ``param_groups`` - a list containing all parameter groups where each parameter group
              is a dict

        """
        return [optimizer.state_dict() for optimizer in self.optimizers]

    def load_state_dict(
        self,
        state_dict: List[
            Dict[str, Union[torch.Tensor, List[Dict[str, Union[torch.Tensor, float, bool, Any]]]]]
        ],
    ) -> None:
        """
        Loads the optimizer state.

        Parameters
        ----------
        state_dict: dict
            Optimizer state. Should be an object returned from a call to ``state_dict()``

        """
        for state, optimizer in zip(state_dict, self.optimizers):
            optimizer.load_state_dict(state)

    def zero_grad(self, set_to_none: bool = False) -> None:
        """
        Sets the gradients of all optimized ``torch.Tensor``s to zero.

        Parameters
        ----------
        set_to_none: bool
            Instead of setting to zero, set the grads to ``None``. This will in general have lower
            memory footprint, and can modestly improve performance. However, it changes certain
            behaviors. For example:

                1. When the user tries to access a gradient and perform manual ops on it, a ``None``
                   attribute or a ``torch.Tensor`` full of ``0``s will behave differently.

                2. If the user requests ``zero_grad(set_to_none=True)`` followed by a backward pass,
                   ``.grad``s are guaranteed to be ``None`` for params that did not receive a
                   gradient.

                3. ``torch.optim`` optimizers have a different behavior if the gradient is ``0`` or
                   ``None`` (in one case it does the step with a gradient of ``0`` and in the other
                   it skips the step altogether).

        """
        for optimizer in self.optimizers:
            optimizer.zero_grad(set_to_none=set_to_none)

    def step(self, closure: Callable[[], torch.Tensor] = None) -> torch.Tensor:
        """
        Performs a single optimization step (parameter update).

        Parameters
        ----------
        closure: function
            A closure that reevaluates the model and returns the loss. Optional for most optimizers.

        Notes
        -----
        Unless otherwise specified, this function should not modify the ``.grad`` field of the
        parameters.

        """
        loss = None

        if closure is not None:
            with torch.enable_grad():
                loss = closure()

        for optimizer in self.optimizers:
            optimizer.step()

        return loss


class HybridLRS(torch.optim.lr_scheduler._LRScheduler):
    """
    Wrapper class around ``lr_scheduler``s to return a dummy optimizer to pass PyTorch Lightning
    checks.

    Modified from the reply in a GitHub Issue thread here:
    https://github.com/Lightning-AI/lightning/issues/3346#issuecomment-1036063687

    Parameters
    ----------
    hybrid_optimizer: HybridOptim
    lr_scheduler: torch.optim.lr_scheduler._LRScheduler
    idx: int
        Index of the optimizer in ``hybrid_optimizer`` the learning rate scheduler ``lr_scheduler``
        is assigned to

    """
    def __init__(
        self,
        hybrid_optimizer: HybridOptim,
        lr_scheduler: torch.optim.lr_scheduler._LRScheduler,
        optimizer_idx: int,
    ) -> None:
        self.optimizer = hybrid_optimizer
        self.lr_scheduler = lr_scheduler
        self.idx = optimizer_idx

    def __getattribute__(self, __name: str) -> Any:
        """
        If the attribute name is one of ``optimizer``, ``idx``, or ``lr_scheduler``, return this
        class's attribute with the same name, else return the ``lr_scheduler``'s attribute with
        that name.

        """
        if __name in {'optimizer', 'lr_scheduler', 'idx'}:
            return super().__getattribute__(__name)
        else:
            return self.lr_scheduler.__getattribute__(__name)
```
